Implementation/64348 formula editor for calculated values

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/64348

# What are you trying to accomplish?

* Use pattern input component in formula field to offer suggestions when editing it.
* Allow inserting other project attributes via dropdown. They will appear as tokens within the formula to make clear that these are "variables".
* Supported project custom field formats:
  * integer
  * float
  * calculated value
* Allow inserting mathematical operators via dropdown, but make them appear as regular text instead of tokens.

- [x] Ensure and validate that only visible custom fields are offered for the formula.
- [x] Make sure that a calculated value is not suggested for its own formula (easiest form of circular reference).
- [x] Fix pattern input bugs
  - [ ] ~Cursor jumps to the front after selection a suggestion from the list (Firefox only).~ Will be fixed separately, [see bug ticket](https://community.openproject.org/wp/66089).
  - [x] Suggestion should open when you edit a pill, so that you can select another custom field instead.

## Screenshots
![image](https://github.com/user-attachments/assets/910560a3-d22e-4e1e-8376-4df546d003fc)

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#64348] Use pattern input for formulas

EinLama committed:
[#64348] suggest operators last

EinLama committed:
[#64348] make operators a constant, i18n

EinLama committed:
[#64348] allow pattern input to insert selection as text

This was already possible with the text template. Now, even existing
suggestions can be configured to behave like text.

EinLama committed:
[#64348] Avoid circular reference, only offer visible fields

EinLama committed:
[#64348] validate formula syntax by building AST

EinLama committed:
[#64348] Modulo operator for Calculated Values

EinLama committed:
[#64348] Fail validation if unknown custom field is in formula

EinLama committed:
[#64348] Only allow patternized variables in formula string

Good: {{cf_21}}
Bad: cf_21

EinLama committed:
[#64348] Fixup specs for Calculated Values

EinLama committed:
[#64348] Spec for #usable_custom_field_references_for_formula

EinLama committed:
[#64348] Adjust documentation

EinLama committed:
[#64348] Refine error message for invalid custom fields

EinLama commented:
☝️the version mismatch is not an issue as the feature is hidden behind a feature flag for now.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -33,12 +33,17 @@
   extend ActiveSupport::Concern
 
+  # Mathematical operators that are allowed in the formula.
+  OPERATORS = %w[+ - * / % ( )].freeze
```
Are we sure `%` will be understandable, given that in Excel and Numbers it means percent, so will not work as `10 % 3`, but will work as `10 + 3%`

EinLama replied:
Thank you for raising this question. It showed that my assumption of what the `%` should mean was not in line with the assumption of others.

As discussed, will update this to work as both percentage and modulo 👍 The `%` will be removed from the suggested operators list, but you can still type it manually.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -33,12 +33,17 @@
   extend ActiveSupport::Concern
 
+  # Mathematical operators that are allowed in the formula.
+  OPERATORS = %w[+ - * / % ( )].freeze
```
Also probably better to use a bit longer/more specific name, as the module is included

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
+
+      if surplus_cfs.any?
+        custom_field_names = CustomField.where(id: surplus_cfs).pluck(:name)
+        errors.add(:formula, :invalid_custom_fields, custom_fields: custom_field_names.join(", "))
```
Would probably be nicer to the user to explain why aren't they allowed, like wrong format, not visible, …

EinLama replied:
I gave this some consideration and attempted an implementation. While doing so, I realized that to implement this properly, we would have to either cache results of previous validation checks, or we would have to perform certain queries and operations again. For example, we would have to check which of the surplus CFs do not have an allowed field format and add an error for them.

After that, we would have to find all fields causing a circular reference and add an error for them.

Same for fields not visible to the user, and so on and so on.

What I am trying to get at is: this would complicate the validation quite a lot. However, if all goes well, users will never add these fields to a formula, since it is never suggested in the first place. So while being nice and helpful, it hopefully will not happen very often. Of course when changing custom fields, a field that was suggested earlier might become invalid after a while. But even that is hopefully a case that does not happen that often.

Therefore, I would keep it vague for the first iteration and observe what the users do with this feature. What do you think?

toy replied:
Sure, I agree, let it be vague. I would only suggest then to change the key from `:invalid_custom_fields` to something like `:not_allowed_custom_fields` or `:not_allowed_referenced_custom_fields`

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
+    def usable_custom_field_references_for_formula
+      scope = type == "ProjectCustomField" ? ProjectCustomField : CustomField
+
+      scope
```
Will this work:
```suggestion
      self.class
```

EinLama replied:
It does! Thank you 👍

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
       formula ? formula.fetch("formula", "") : ""
     end
 
+    def usable_custom_field_references_for_formula
```
Probably better to exclude those which will lead to recursion

EinLama replied:
Good point, working on it.

EinLama replied:
I have pushed a check for circular references. I implemented it in Ruby since it's (IMO) much simpler than adding a CTE for checking it directly on the database and performance for this particular validation is not yet a big concern. This might change when this feature sees heavy use in the future.

toy replied:
I don't think we should worry about performance of this, as validation is happening only when updating the custom field, so should not be an often repeating action

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -49,15 +54,12 @@ def validate_formula
 
-      # TODO: consider differentiating between a formula that contains missing variables, invalid
-      #       syntax, or mathematical errors.
+      true
```
Not needed for this method, as return value is not used

EinLama replied:
I only included the `true` to return _something_ since RuboCop would otherwise complain about the unnecessary return in the above `unless` block.

And I did not want to remove _that_ as we might add lines to this method in the future - and we would have to remember to go back to the above `unless` block and add the return back in.

But you know what? This entire discussion shows how twisted my thought process was here. We don't need these redundant returns at the moment. You're right. Removed!

toy replied:
If there would be `nil` instead of `true` I would not complain.
Also probably a controversial idea:

```rb
case
when formula_string.blank?
  errors.add(:formula, :blank)
when !formula_contains_only_allowed_characters?
  errors.add(:formula, :invalid_characters)
when !valid_formula_syntax?
  errors.add(:formula, :invalid)
end
```

EinLama replied:
That's nice and more compact. I took that idea and transformed it into an if-construct now since RuboCop doesn't like a case-statement without `else`:

```ruby
      if formula_string.blank?
        errors.add(:formula, :blank)
      elsif !formula_contains_only_allowed_characters?
        errors.add(:formula, :invalid_characters)
      elsif !valid_formula_syntax?
        errors.add(:formula, :invalid)
      end
```

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
+      # We can only validate used custom fields from a high-level perspective, since at this point in validation,
+      # we do not have a project context to check against. So we cannot check if the custom fields are actually
+      # enabled for a project and visible to a non-admin user.
+      formula_cfs = CustomField.where(id: cf_ids_used_in_formula(formula_string)).pluck(:id)
```
I think it is better to just use result of `cf_ids_used_in_formula(formula_string)`, as ids of non existing custom fields are also wrong
```suggestion
      formula_cfs = cf_ids_used_in_formula(formula_string)
```

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
     private
 
+    def calculator
+      @calculator ||= Dentaku::Calculator.new(case_sensitive: true)
```
Why do you want to cache it?

EinLama replied:
I thought it might be useful for when we use the calculator object multiple times in a short time. E.g., when we first validate the formula - and then calculate the numerical value. But my assumption was not valid: at this point in time, there is only one method call happening. So there is no use in caching this instance. Removed!

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
+    # For a formula like `2 + {{cf_12}} + {{cf_4}}` it returns `[12, 4]`.
     def cf_ids_used_in_formula(formula_str)
-      formula_str.scan(/\bcf_(\d+)\b/).flatten.map(&:to_i)
+      formula_str.scan(/\{\{cf_(\d+)}}/).flatten.compact.map(&:to_i)
```
I don't think compact is needed here, also `flatten` can be removed by using lookaheads:
```suggestion
      formula_str.scan(/(?<=\{\{cf_)\d+(?=}})/).map(&:to_i)
```

EinLama replied:
Oh, that's great! TIL thank you.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +75,75 @@ def formula_string
-      allowed_chars = %w[+ - / * ( )] + [" "]
-      allowed_tokens = /\A(cf_\d+|\d+\.?\d*|\.\d+)\z/
+      allowed_chars = OPERATORS + [" "]
+      allowed_tokens = /\A(\{\{cf_\d+}}|\d+\.?\d*|\.\d+)\z/
```
Instead of `\d+\.?\d*`, better `\d+(\.\d+)?`

EinLama replied:
With that, the validation would fail at this level for a formula containing trailing decimal points, like: `1.5 + 1. - 3.25`

We want the validation to fail, but at a later point in time. Per se, the `.` is allowed, but only when there's a number following it. The fail message `formula contains invalid characters.` is slightly misleading.

The incorrect placement of the floating point will be validated later when Dentaku fails to parse it, with a slightly better fitting - although still not great - validation message: `formula is invalid`.

Both solutions aren't perfect, but I personally like the current behavior a bit better.

We can discuss whether it is worth it to add another level of validation with a customized fail message exactly for this case. For now, I considered the current validation good enough™

I am not sure yet how users will use the formula.

toy replied:
To handle than case then it can be `\d+(\.\d*)?`, but maybe it is better to make it then even more relaxed allowing any combination of numbers and dots:
`/\A(\{\{cf_\d+}}|[\d.]+)\z/`

EinLama replied:
I like the simplicity 👍

toy reviewed (commented):
I did not look into specs for now

EinLama committed:
[#64348] Use a more explicit name for formula operators

EinLama committed:
[#64348] Remove redundant returns

EinLama committed:
[#64348] Use self.class for scoping

EinLama committed:
[#64348] Do not cache calculator instance

EinLama committed:
[#64348] Take formula cfs directly from formula

EinLama committed:
[#64348] Use lookaheads/-behinds to get rid of flatten and compact

EinLama committed:
[#64348] Remove % from the suggestion list

It can be used as either modulo or for percentage calculations. Removing
it from the suggestion list for mathematical operators as that might be
confusing. You can still type % manually, though. Added specs for these
cases.

EinLama committed:
[#64348] Simplify character validation

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +72,72 @@ def formula_string
+        .where(field_format: FIELD_FORMATS_FOR_FORMULA)
+        # Disallow the current custom field to avoid circular references
+        .where.not(id: id)
+        .visible
```
`visible` is not available for all types of custom fields, so not for `CustomField` or `WorkPackageCustomField`

EinLama replied:
Good catch. I hardcoded ProjectCustomField into this method since the field format is only available as a project attribute right now. When adding support for work package custom fields, we have to change this and other places, there is no need to prepare it at this stage.

EinLama committed:
[#64348] Use compact if/elsif instead of early returns

EinLama committed:
[#64348] Avoid circular references

EinLama committed:
[#64348] Remove TODO comment

To implement this properly, we would have to either cache results of
previous validation checks, or we would have to perform certain queries
and operations again. For example, we would have to check which of the
surplus CFs do not have an allowed field format and add an error for
them.

After that, we would have to find all fields causing a circular
reference and add an error for them.

Same for fields not visible to the user, and so on and so on.

What I am trying to get at is: This would complicate matters quite a
lot. However, if all goes well, users will never add these fields to a
formula, since it is never suggested in the first place. So while being
nice and helpful, it hopefully will not happen very often. Therefore, I
would keep it vague for the first iteration and observe what the users
do with this feature.

EinLama commented:
I have added two more TODO-items about fixing frontend bugs to the list in the PR description. Will tackle these now and ping you when I am done :)

EinLama committed:
[#64348] Allow editing and replacing existing tokens within formula

Had to mark label_with_context as optional as we do not use it for
calculated values.

toy committed:
[#64348] add formula_referenced_custom_field_ids method

toy commented on app/forms/custom_fields/calculated_values/details_form.rb:
```diff
@@ -75,6 +76,28 @@ class DetailsForm < ApplicationForm
+      private
+
+      def formula_suggestions
+        operators = CustomField::CalculatedValue::MATH_OPERATORS_FOR_FORMULA
+                      # Hide % from the suggestions as it can be used as either modulo or percentage.
+                      .reject { it == "%" }
```
I think using `-` is more idiomatic:
`CustomField::CalculatedValue::MATH_OPERATORS_FOR_FORMULA - %w[%]`

EinLama replied:
For most use cases I agree. But given the map call in the following line, I like `reject` here better. It looks clearer to me. It also allows us to place the explaining comment directly above the line where it's needed, without any additional logic in the way.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -33,31 +33,24 @@
   included do
-    validate :validate_formula
+    validate :validate_formula, if: :field_format_calculated_value?
+    validate :validate_referenced_custom_fields_allowed, if: :field_format_calculated_value?
```
I think name can be shortened to `validate_referenced_custom_fields` to express that referenced fields are checked for all violations. Also on second thought it feels like it can also be called from `validate_formula` instead, as it makes sense to check referenced fields only if formula is valid

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+
+      visible_cfs.reject do |custom_field|
+        has_circular_reference?(custom_field, id)
+      end
```
Not sharing `visited` between calls to `has_circular_reference?` is making it less efficient, there may be different ways to resolve it

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+      visited.add(custom_field.id)
+
+      if custom_field.field_format_calculated_value? && custom_field.formula
+        referenced_custom_fields = custom_field.formula.fetch("referenced_custom_fields", [])
```
I've added `formula_referenced_custom_field_ids`, you can cherry pick 2f1eafa9805504e5332bc5d2c0de024c25794829

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
 
       formula_string.split(Regexp.union(allowed_chars)).reject(&:empty?).all? do |token|
         token.match?(allowed_tokens)
       end
```
Can also be:
```suggestion
      formula_string.split(Regexp.union(allowed_chars)).all? do |token|
        token.empty? || token.match?(allowed_tokens)
      end
```

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+    # For a formula like `2 + {{cf_12}} + {{cf_4}}` it returns `[12, 4]`.
     def cf_ids_used_in_formula(formula_str)
-      formula_str.scan(/\bcf_(\d+)\b/).flatten.map(&:to_i)
+      formula_str.scan(/(?<=\{\{cf_)\d+(?=}})/).map(&:to_i)
```
Should this have `.uniq` at the end?

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+    end
+
+    def has_circular_reference?(custom_field, original_id, visited = Set.new)
+      return true if visited.include?(custom_field.id)
```
This will lead to false positives, for example if calculated value references constant cf_1 and calculated cf_2, while cf_2 also references cf_1, result will be true

EinLama replied:
Hmm, that should not result in `has_circular_reference?` returning true.

The method checks it from the view of the calculated value `cv`. In your example, it will see if either `cf_1` or `cf_2` contain a reference to `cv`, which they do not. So both pass this check and can be used within the formula.

To illustrate this, I wrote this spec, which is green on my branch and confirms my theory. Please have a look if I understood it correctly:

```ruby
    context "with Ivans circular reference example" do
      # for example, if CV references constant cf_1 and calculated cf_2, while cf_2 also references cf_1
      let!(:constant_cf1) { create(:project_custom_field, :integer, default_value: 1, is_for_all: true) }
      let!(:calculated_cf2) do
        create(:calculated_value_project_custom_field, formula: "1 + {{cf_#{constant_cf1.id}}}", is_for_all: true)
      end

      it "does not lead to a false positive" do
        subject.formula = "{{cf_#{constant_cf1.id}}} + {{cf_#{calculated_cf2.id}}}"
        subject.save(validate: false)

        expect(subject.usable_custom_field_references_for_formula).to include(constant_cf1, calculated_cf2)
      end
    end
```

EinLama replied:
Maybe it becomes easier with the refactor of the method: https://github.com/opf/openproject/pull/19412#discussion_r2240002016

toy replied:
The false positives show up one level deeper:

```ruby
context "with Ivans circular reference example" do
  # problem comes up when some field is referenced by two different custom fields
  let!(:constant_cf1) { create(:project_custom_field, :integer, default_value: 1, is_for_all: true) }
  let!(:calculated_cf2) do
    create(:calculated_value_project_custom_field, formula: "{{cf_#{constant_cf1.id}}}", is_for_all: true)
  end
  let!(:calculated_cf3) do
    create(:calculated_value_project_custom_field, formula: "{{cf_#{constant_cf1.id}}} + {{cf_#{calculated_cf2.id}}}", is_for_all: true)
  end

  it "does not lead to a false positive" do
    subject.formula = "{{cf_#{calculated_cf3.id}}}"
    subject.save(validate: false)

    expect(subject.usable_custom_field_references_for_formula).to include(constant_cf1, calculated_cf2, calculated_cf3)
  end
end
```

toy replied:
I think following will resolve both recursion false positive and allows to not recheck custom fields (if cache is shared in the `visible_cfs.reject do |custom_field|` cycle):
```ruby
def formula_references_id?(original_id, cache = {})
  cache.fetch(id) do
    cache[id] = if field_format_calculated_value?
      referenced_custom_fields = formula_referenced_custom_field_ids

      next true if referenced_custom_fields.include?(original_id)

      ProjectCustomField.where(id: referenced_custom_fields).any? do |referenced_field|
        referenced_field.formula_references_id?(original_id, cache)
      end
    else
      false
    end
  end
end
```

EinLama replied:
Ah, now I see. Thank you! I adjusted a detail to prevent an infinite loop from happening in an edge-case. Now, all specs are green again. I also included your example into the specs 👍

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+      formula_string.gsub(/\{\{(cf_\d+)}}/, '\1')
+    end
+
+    def has_circular_reference?(custom_field, original_id, visited = Set.new)
```
I find the method name confusing, as it suggests that there is already a circular reference, but it tells if original_id is referenced by custom_field or its references

EinLama replied:
You are right. I have renamed it to `formula_references_id?` and also changed it so that `custom_field` is not passed into the method, but instead you call `custom_field.formula_references_id?`. That reads better. I adjusted the specs for that, too.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+    def has_circular_reference?(custom_field, original_id, visited = Set.new)
+      return true if visited.include?(custom_field.id)
+
+      visited.add(custom_field.id)
```
Can be combined with `include?` check by using `!visited.add?(custom_field.id)`

EinLama replied:
Oh nice, didn't know about this method 💡

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +66,92 @@ def formula_string
+        return true if referenced_custom_fields.include?(original_id)
+
+        referenced_custom_fields.any? do |ref_id|
+          referenced_field = ProjectCustomField.find_by(id: ref_id)
```
Probably more efficient to query all referenced in one go:
```suggestion
        ProjectCustomField.where(id: referenced_custom_fields).any? do |referenced_field|
```

toy reviewed (changes requested):
The recursion check has a problem, otherwise small things and nitpicks

EinLama committed:
[#64348] Review feedback: increase code efficiency

EinLama committed:
[#64348] use #formula_referenced_custom_field_ids

EinLama committed:
[#64348] mv has_circular_reference? formula_references_id?

EinLama committed:
[#64348] Refactor #formula_references_id?

It is now a proper instance method that is called on the current custom
field instead of weirdly accepting another custom field as first
argument.

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +67,93 @@ def formula_string
+      # We can only validate used custom fields from a high-level perspective, since at this point in validation,
+      # we do not have a project context to check against. So we cannot check if the custom fields are actually
+      # enabled for a project and visible to a non-admin user.
+      formula_cfs = cf_ids_used_in_formula(formula_string)
```
Can use `formula_referenced_custom_field_ids`

toy commented on spec/models/custom_field/calculated_value_spec.rb:
```diff
@@ -56,13 +149,193 @@
+    end
+  end
+
+  describe "#formula_contains_reference_to_id?" do
```
```suggestion
  describe "#formula_references_id?" do
```

toy commented on spec/models/custom_field/calculated_value_spec.rb:
```diff
@@ -56,13 +149,193 @@
+
+    context "when checking a non-calculated value custom field" do
+      it "returns false for integer custom field" do
+        expect(int_field.formula_references_id?(subject.id)).to be false
```
Just an idea, really not sure if it is better:
```suggestion
        expect(int_field).to be_formula_references_id(subject.id)
```

EinLama replied:
I thought about that, but with the method name being what it is, it does not read like proper English and feels off.

EinLama committed:
[#64348] Fix formula reference checking

The previous version lead to false positives

EinLama committed:
[#64348] Names

toy commented on app/models/custom_field/calculated_value.rb:
```diff
@@ -73,29 +67,94 @@ def formula_string
+
+    def formula_references_id?(original_id, cache = {})
+      cache.fetch(id) do
+        if field_format_calculated_value?
```
`fetch` doesn't assign value, so storing explicitly is needed
```suggestion
        cache[id] = if field_format_calculated_value?
```

EinLama replied:
Sorry about that. I removed that line during testing something and failed to put it back. Fixed.

EinLama committed:
[#64348] Assign value in fetch

toy reviewed (approved):
Very good, I didn't notice anything else
